File: GasProject Server/Sources/World/Objects/Turfs/Airlock.cpp
#include "Airlock.hpp"

#include <functional>

#include "Server.hpp"
#include "World/World.hpp"
#include "Network/Differences.hpp"

Airlock::Airlock() {
    name = "airlock";
    sprite = "airlock";
    density = true;
    opened = false;
    locked = false;
}

void Airlock::Update(sf::Time timeElapsed) {
	Turf::Update(timeElapsed);
	closeTimer.Update(timeElapsed);
}

bool Airlock::InteractedBy(Object *object) {
	if (!Turf::InteractedBy(object))
		return false;
    Activate();
	return true;
}

void Airlock::Activate() {
	// Blink if locked 
    if (locked) {
        if (!opened) {
            PlayAnimation("airlock_closed_animation");
        }
        return;
    }

	if (opened) {
		if (!PlayAnimation("airlock_closing"))
			return;
		SetSprite("airlock");
		opened = false;
		density = true;
	} else {
		if (!PlayAnimation("airlock_opening", std::bind(&Airlock::animationOpeningCallback, this)))
			return;
		SetSprite("airlock_opened");
		closeTimer.Start(AUTOCLOSE_TIME, std::bind(&Airlock::autocloseCallback, this));
	}
}

void Airlock::Lock() {
    locked = true;
}

void Airlock::Unlock() {
    locked = false;
}

void Airlock::animationOpeningCallback() {
	opened = true;
	density = false;
}

void Airlock::autocloseCallback() {
	if (opened)
		Activate();
}

File: Script/Objects/Turfs/Airlock.py
from Objects.Turf import Turf

class Airlock(Turf):
	def __init__(self):
		super().__init__()
		self.name = "Airlock"
		self.sprite = "airlock"
		self.density = True

		self.opened = False
		self.locked = False
		
	def InteractedBy(self, object):
		if not self.IsCloseTo(object):
			return False
		self.Activate()
		return True

	def Activate(self):
		if self.locked:
			if not self.opened:
				self.PlayAnimation("airlock_closed_animation")
			return

		if self.opened:
			if not self.PlayAnimation("airlock_closing"):
				return
			self.sprite = "airlock"
			self.opened = False
			self.density = True
		else:
			if not self.PlayAnimation("airlock_opening", lambda: self.__animationOpeningCallback()):
				return
			self.sprite = "airlock_opened"

	def __animationOpeningCallback(self):
		self.opened = True
		self.density = False

	def __autocloseCallback(self):
		if self.opened:
			self.Activate()

